src/Npgsql/Internal/Converters/Temporal/DateTimeConverterResolver.cs

```diff
@@ -0,0 +1,143 @@
+using System;
+using System.Collections.Generic;
+using Npgsql.Internal.Postgres;
+using Npgsql.Properties;
+using NpgsqlTypes;
+
+// ReSharper disable once CheckNamespace
+namespace Npgsql.Internal.Converters;
+
+sealed class DateTimeConverterResolver<T> : PgConverterResolver<T>
+{
+    readonly PgSerializerOptions _options;
+    readonly Func<DateTimeConverterResolver<T>, T?, PgTypeId?, PgConverterResolution?> _resolver;
+    readonly Func<PgTypeId, PgConverter> _factory;
+    readonly PgTypeId _timestampTz;
+    PgConverter? _timestampTzConverter;
+    readonly PgTypeId _timestamp;
+    PgConverter? _timestampConverter;
+    readonly bool _dateTimeInfinityConversions;
+
+    internal DateTimeConverterResolver(PgSerializerOptions options, Func<DateTimeConverterResolver<T>, T?, PgTypeId?, PgConverterResolution?> resolver, Func<PgTypeId, PgConverter> factory, PgTypeId timestampTz, PgTypeId timestamp, bool dateTimeInfinityConversions)
+    {
+        _options = options;
+        _resolver = resolver;
+        _factory = factory;
+        _timestampTz = timestampTz;
+        _timestamp = timestamp;
+        _dateTimeInfinityConversions = dateTimeInfinityConversions;
+    }
+
+    public override PgConverterResolution GetDefault(PgTypeId? pgTypeId)
+    {
+        if (pgTypeId == _timestampTz)
+            return new(_timestampTzConverter ??= _factory(_timestampTz), _timestampTz);
+        if (pgTypeId is null || pgTypeId == _timestamp)
```

Comment:
That is to support a 'best-default', which purely exists for BinaryExporter where we don't have a pgtypeid on reads. I don't have this particular defaulting at all in Slon, as it's definitely weird...

Reply:
>Were we basically doing the same thing before

So before we used to do ResolveByClrType which would then try to do clr type to datatypename and end up here 
https://github.com/npgsql/npgsql/blob/main/src/Npgsql/TypeMapping/BuiltInTypeMappingResolver.cs#L153

(Weirdly enough it's in there and also added in a LegacyTimestampBehavior branch at the end of the method)

It seems like it would have worked, but mostly by accident?

Reply:
Yeah, I definitely didn't have the binary export case well in mind when working on all of this.

Anyway, it all looks good for the current state of affairs, hopefully at some point we can improve COPY somehow.